### PR TITLE
Community Wellness badge

### LINKED ISSUE #17310: Community Wellness Badge
We would like to build a community wellness badge styled after the writing streak badge.  

Badge name: Community Wellness
Type: Streak of badges based on activity
Level 1: posting at least 2 constructive comments in a single week
Level: 2: posting at least 2 constructive comments per week for 2 straight weeks
Level: 3: posting at least 2 constructive comments per week for 4 straight weeks
Level: 4: posting at least 2 constructive comments per week for 8 straight weeks
Level: 5: posting at least 2 constructive comments per week for 16 straight weeks
Level: 6: posting at least 2 constructive comments per week for 32 straight weeks

“constructive” is defined as not flagged as negative by mods. The mods will be actively monitoring for productive conversation and flagging comments that aren’t constructive.

The DP team will provide graphics.

### PR DESCRIPTION
**Pull request by fdocr**
## What type of PR is this? (check all applicable)

- [x] Feature

## Description

Implements the Community Wellness Badge behind a Feature Flag

## Related Tickets & Documents

- Relates to #17310

## QA Instructions, Screenshots, Recordings

Since this is a feature that works with data that exists over time it may be tricky to test locally (check out the specs in place). However, if feeling confident about trying it out locally you can:

1. Boot app locally
2. Create comments on existing posts
3. Update the comment dates (and their corresponding reactions) using the console
   - You need at least two comments per week
   - Comments must be at least 3 days old (first week award)
   - Reaction dates may have to be updated like this as well ([spec helper takes this into account](https://github.com/forem/forem/pull/17394/files#diff-adacbae246695183a46a623a417898510a8e7282dda432c18775703e60c9798eR12))
5. Run the service using the console
   - `Badges::CommunityWellnessBadge.call`
6. Check out Ahoy tracking to see if the Badge was attempted to be awarded on your user

If you add a negative reaction on the comments (mod/trusted user) the comment shouldn't count for the badge

### UI accessibility concerns?

None - no new HTML was introduced

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] I will share this change internally with the appropriate teams

## [optional] Are there any post deployment tasks we need to perform?

Keep an eye out for the next execution of the badge service (~00:45 UTC) and double check results/performance

## [optional] What gif best describes this PR or how it makes you feel?

riding unicorn through data

**Above: The new service parsing through the `comments` table**

### CODE REVIEW

File: config/schedule.yml [hunk 1]
```diff
@@ -78,6 +78,14 @@ award_contributor_badges_from_github:
...
     - award_contributor_from_github
     - ""
+award_community_wellness_badges:
+  description: "Awards 'Community Wellness' badges to users (runs at minute 25 past every 12th hour)"
+  cron: "25 */12 * * *"
+  class: "BadgeAchievements::BadgeAwardWorker"
+  args:
+    - ""
+    - award_community_wellness
+    - ""
```
Comment (fdocr) [inline config/schedule.yml:88]: I **think** this job can run maybe twice a day? This was a quick cron setting put together with crontab.guru, so I think there's likely better ways to schedule this.

The queries will check for 1 week windows into the past, so if I make 2 unflagged comments in the "past week", then a week goes by and they fall into the "second week ago" (2.weeks.ago - 1.weeks.ago) we have the "first week ago" (1.weeks.ago - 0.weeks.ago/Time.current) to look for the other 2 unflagged comments. So, twice a day or even just once per day might be enough to avoid hitting the DB too hard?

I also chose to run at the minute 25, since other jobs run at the 0/12/20 minute on the hour, to try avoid putting too much pressure on the DB at once. Does this make sense?

**Comment by jeremyf** [comment]
@fdocr I wrote up a SQL that I believe can create a single query to get you the relevant information:

```sql
SELECT user_id,
	 COUNT(user_id),
	 (trunc((extract(epoch FROM (current_timestamp - created_at))) / 604800)) AS number_of_weeks
FROM comments
INNER JOIN
      (SELECT reactable_id
      FROM reactions
      WHERE reactable_type = 'Comment'
      	    AND category NOT IN ('thumbsdown', 'vomit')
	    AND created_at > (now() - interval '224' day)
	    GROUP BY reactable_id) AS positve_reactions
	    ON comments.id = positve_reactions.reactable_id
WHERE user_id = :user_id
      AND created_at > (now() - interval '224' day)
GROUP BY user_id, number_of_weeks
```


| user_id  | count | week |
|----------|-------|------|
| :user_id |    19 |  0.0 |
| :user_id |    59 |  1.0 |
| :user_id |    25 |  2.0 |
| :user_id |    29 |  3.0 |
| :user_id |    18 |  4.0 |
| :user_id |     4 |  5.0 |
| :user_id |     4 |  6.0 |
| :user_id |    14 |  8.0 |
| :user_id |     4 |  9.0 |
| :user_id |     4 | 10.0 |
| :user_id |    17 | 11.0 |
| :user_id |    17 | 12.0 |
| :user_id |    28 | 13.0 |
| :user_id |    19 | 14.0 |
| :user_id |    21 | 15.0 |
| :user_id |     6 | 16.0 |
| :user_id |    18 | 17.0 |
| :user_id |     1 | 19.0 |
| :user_id |     3 | 20.0 |
| :user_id |     1 | 21.0 |
| :user_id |     5 | 22.0 |
| :user_id |     6 | 23.0 |
| :user_id |     6 | 24.0 |
| :user_id |    16 | 25.0 |
| :user_id |     9 | 26.0 |
| :user_id |     5 | 27.0 |
| :user_id |    37 | 28.0 |
| :user_id |    10 | 29.0 |
| :user_id |     7 | 30.0 |
| :user_id |     2 | 31.0 |

These queries would be an alternate approach to the iterative methodology.  Also, I believe there's an off by one error as well (0 to 31 instead of 1 to 32).

I'd want @jgaskins to review the SQL for correctness as well as performance.

**Comment by fdocr** [comment]
This is very interesting thanks for sharing it @jeremyf! A couple of questions:

1. Does each `row` in the result represent a comment that was posted X weeks ago? (X = the value in this column)
   - Meaning we'll have to aggregate this result to realize how many comments (rows) exist on week 1, week 2, week 3, etc., right? Not necessarily an issue per-se but trying to understand how to process these results after its executed
3. The query is restricted with `user_id = 1`. Can we get rid of that or have it join with the result of all users with at least 2 comments in the last week?
   - Unrestricted by `user_id` the result set is quite large but it didn't timeout on Blazer for me, so joining with those users that have at least 2 comments in the last week might make it more manageable

**Comment by jeremyf** [comment]
> Does each row in the result represent a comment that was posted X weeks ago? (X = the value in number_of_weeks column)

I added some comments to the SQL in Blazer and renamed a variable.

Each row represents a user, the number of "positive reacted comments" they got, for the given "weeks ago".  So row one is user 1 had 17 comments this week that had at least two positive reactions each.

To process, we'd group the results by the user id and then inspect the weeks.

> Can we get rid of that or have it join with the result of all users with at least 2 comments in the last week?

Yes.  I'll need to work on that just a bit.

**Comment by jgaskins** [comment]
@fdocr Each row is pre-aggregated comment stats for the week, so application code should be able to use the results as-is without in-memory aggregation. However, for your point 2, if we collect all `user_id`s, we would then have to aggregate them for the data to be useful. If we want to get results for a collection of users, it might (or might not) be preferable to have Postgres aggregate the rows for us using something like `json_agg`. Depends on how much data we're talking about and how often this query needs to be run.

@jeremyf This query looks pretty accurate at first glance (query plan below), but it does a couple O(n) filter operations on DEV's dataset.

<details><summary>Full query plan</summary>

```
                                                                                          QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=92296.42..92296.70 rows=32 width=40) (actual time=1348.825..1348.899 rows=29 loops=1)
   Group Key: comments.user_id, (trunc((date_part('epoch'::text, (CURRENT_TIMESTAMP - (comments.created_at)::timestamp with time zone)) / '604800'::double precision)))
   ->  Sort  (cost=92296.42..92296.43 rows=32 width=24) (actual time=1348.814..1348.836 rows=352 loops=1)
         Sort Key: (trunc((date_part('epoch'::text, (CURRENT_TIMESTAMP - (comments.created_at)::timestamp with time zone)) / '604800'::double precision)))
         Sort Method: quicksort  Memory: 52kB
         ->  Hash Join  (cost=90254.10..92296.26 rows=32 width=24) (actual time=1347.100..1348.721 rows=352 loops=1)
               Hash Cond: (comments.id = positve_reactions.reactable_id)
               ->  Bitmap Heap Scan on comments  (cost=850.68..2892.17 rows=1092 width=24) (actual time=20.867..22.289 rows=414 loops=1)
                     Recheck Cond: ((user_id = 1) AND (created_at > (now() - '224 days'::interval)))
                     Heap Blocks: exact=398
                     ->  BitmapAnd  (cost=850.68..850.68 rows=1092 width=0) (actual time=20.784..20.785 rows=0 loops=1)
                           ->  Bitmap Index Scan on index_comments_on_user_id  (cost=0.00..87.98 rows=9262 width=0) (actual time=2.362..2.363 rows=8853 loops=1)
                                 Index Cond: (user_id = 1)
                           ->  Bitmap Index Scan on index_comments_on_created_at  (cost=0.00..762.54 rows=93636 width=0) (actual time=17.962..17.962 rows=95686 loops=1)
                                 Index Cond: (created_at > (now() - '224 days'::interval))
               ->  Hash  (cost=89322.54..89322.54 rows=23108 width=16) (actual time=1326.206..1326.207 rows=45829 loops=1)
                     Buckets: 65536 (originally 32768)  Batches: 1 (originally 1)  Memory Usage: 2661kB
                     ->  Subquery Scan on positve_reactions  (cost=89010.58..89322.54 rows=23108 width=16) (actual time=1290.895..1316.617 rows=45829 loops=1)
                           ->  HashAggregate  (cost=89010.58..89253.22 rows=23108 width=16) (actual time=1290.894..1312.582 rows=45829 loops=1)
                                 Group Key: reactions.reactable_id
                                 Filter: (count(reactions.id) > 1)
                                 Batches: 1  Memory Usage: 11281kB
                                 Rows Removed by Filter: 53399
                                 ->  Index Scan using index_reactions_on_created_at on reactions  (cost=0.09..88616.80 rows=262525 width=16) (actual time=0.062..1222.065 rows=184038 loops=1)
                                       Index Cond: (created_at > (now() - '224 days'::interval))
                                       Filter: (((category)::text <> ALL ('{thumbsdown,vomit}'::text[])) AND ((reactable_type)::text = 'Comment'::text))
                                       Rows Removed by Filter: 1200380
 Planning Time: 0.577 ms
 Execution Time: 1351.033 ms
```

</details>

The big thing to look for is `Rows Removed by Filter`. This means that the row didn't match your query but Postgres still had to look at it anyway. We often think about performance in terms of how many rows are returned, and that's certainly part of it and definitely the most visible part, but how many rows Postgres has to touch _at all_ during the query often has the larger yield.

One of these filters is not amazing but still very tolerable, at 53k rows removed from the results:

```
                           ->  HashAggregate  (cost=89010.58..89253.22 rows=23108 width=16) (actual time=1290.894..1312.582 rows=45829 loops=1)
                                 Group Key: reactions.reactable_id
                                 Filter: (count(reactions.id) > 1)
                                 Batches: 1  Memory Usage: 11281kB
                                 Rows Removed by Filter: 53399
```

It _might_ still be significant because it's filtering on aggregated data (aggregations are slower than returning unprocessed table data), but this part is likely be more intense:

```
                                 ->  Index Scan using index_reactions_on_created_at on reactions  (cost=0.09..88616.80 rows=262525 width=16) (actual time=0.062..1222.065 rows=184038 loops=1)
                                       Index Cond: (created_at > (now() - '224 days'::interval))
                                       Filter: (((category)::text <> ALL ('{thumbsdown,vomit}'::text[])) AND ((reactable_type)::text = 'Comment'::text))
                                       Rows Removed by Filter: 1200380
```

1.2 Million rows filtered. For some reason, it's using the index on `created_at`, but not on `category` or `reactable_type`. Both columns are indexed but neither of those indexes are being used. I tried changing the `NOT IN` comparator to two `!=` conditions and it didn't change anything meaningfully. I even tried creating a temporary index (and running a `VACUUM ANALYZE` so the query planner can get the new heuristics) and it _still_ didn't change anything.

If this query is run sparingly (a few times per day for statistics gathering, for example), the 1-1.5 seconds it takes to run it may be acceptable for now. If it's going to be run frequently, we're going to want to speed it up by an order of magnitude. If we want to run this for all users, for example, we can't run this 800k times per day. 

@fdocr If your query that removed the filter on `user_id` ran in roughly the same time, we might be able to roll with that. Though, we'll need to keep in mind memory usage of a query that size for both the DB and the app — especially for folks not running on infrastructure anywhere near as beefy as DEV's. Happy to pair on this!

**Comment by jeremyf** [comment]
I forked the query and added: https://dev.to/admin/blazer/queries/672-users-with-comments-this-week-and-their-distribution-of-number-of-comments-by-week-with-2-or-more-positive-reactions

File: app/services/badges/award_community_wellness.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      # subset of users and not large joins.
+      users = User.where(id: multiple_comment_user_ids)
+
+      users.find_each do |user|
+        week_streak = 0
+        (1..REWARD_STREAK_WEEKS.last).each do |week|
+          break unless wellness_goal_x_weeks_ago?(user, week)
+
+          week_streak += 1
+        end
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply app/services/badges/award_community_wellness.rb:21]: This is a really smart optimization, and a fantastic explanation!

File: config/schedule.yml (hunk 1 above)
Comment (joshpuetz) [reply config/schedule.yml:88]: [Opinion] I definitely don't think this badge needs to be run hourly: it's not a big deal if a user doesn't receive their week streak badge **exactly** at the one week mark on the dot.

**Comment by jeremyf** [comment]
Just to make sure we don't lose the query we concocted: https://dev.to/admin/blazer/queries/677-users-with-comment-counts-and-weeks-ago-one-row-per-user

File: app/services/badges/award_community_wellness.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (fdocr) [reply app/services/badges/award_community_wellness.rb]: Thanks @joshpuetz! A condensed SQL query will be more performant in comparison though. Thanks to the help from @jeremyf and @jgaskins to write it I was just able to push an updated approach (still has some pending details to polish).

Greedy algorithms are fun, so maybe a solution like this will be viable in some other time/area of the codebase 😄

File: config/schedule.yml (hunk 1 above)
Comment (fdocr) [reply config/schedule.yml:88]: Sounds great! Right now I'm feeling like running it once per day would be enough then. And we can always revisit this once we understand how it performs in the wild 🙂

**Comment by fdocr** [comment]
Quick update: The service seems to be working as expected now but there's pending things to do before opening for review

- [ ] Actually award the badge
- [ ] Use proper badge texts & assets/design
- [ ] Test coverage for the feature

File: app/queries/comments/community_wellness_query.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        SELECT user_id,
+             /* A comma separated string of "weeks_ago" */
+             array_to_string(array_agg(weeks_ago), ',') AS serialized_weeks_ago,
+             /* A comma separated string of comment counts.  The first value in this string happens on the week that is the first value in serialized_weeks_ago */
+             array_to_string(array_agg(number_of_comments_with_positive_reaction), ',') AS serialized_comment_counts
+        FROM
+        (SELECT user_id,
+             COUNT(user_id) as number_of_comments_with_positive_reaction,
+             /* Get the number of weeks, since today for posts */
+             (trunc((extract(epoch FROM (current_timestamp- created_at))) / 604800) + 1) AS weeks_ago
```
Comment (fdocr) [inline app/queries/comments/community_wellness_query.rb:37]: FYI @jeremyf I added a `+1` to the truncate so indexing starts at `1` instead of `0`. This simplifies the logic in the service that processes these results.

I wasn't sure if `ceil` would work as a replacement here, in the specific case where the result is already a number without decimals (unlikely but possible?).

This one and removing the unserialized result column are the only changes made to the query. Since we're not instantiating ActiveRecord objects we don't gain anything from it, on top of some other details that weren't that helpful even with that in place (results arrived in scientific notation and had to be checked against integers).

Anyways, tests and tweaks will be coming soon before taking out of Draft 🤞

File: config/locales/services/en.yml [hunk 4]
```diff
@@ -27,6 +27,10 @@ en:
           other: Happy %{community} birthday! Can you believe it's been %{count} years already?!
       congrats: Congrats!!!
       thank_you: Thank you so much for your contributions!
+      community_wellness:
+        first: Thank you for participating in constructive conversation! You posted two or more comments this week. Keep it up to continue the streak!
+        longest: 32 weeks! You've achieved the longest streak possible for Community Wellness badge. This makes you eligible for special quests in the future. Keep up the amazing contributions to our community!
+        other: Congrats on achieving a %{weeks} week streak for the Community Wellness Badge! Keep the constructive comments flowing to continue the streak
```
Comment (fdocr) [inline config/locales/services/en.yml:33]: We need a copy review of these texts :)

File: config/schedule.yml (hunk 1 above)
Comment (JennieOcken) [reply config/schedule.yml:88]: Agree on the idea that it only needs to run once a day.

is `"Awards 'Community Wellness' badges to users (runs at minute 25 past every 12th hour)"` visible to users or is it just the name of the cron job?  If the latter, I have no opinion.  If it shows up in the UI it's a little hard to parse.

Comment (fdocr) [reply config/schedule.yml:88]: > is `"Awards 'Community Wellness' badges to users (runs at minute 25 past every 12th hour)"` visible to users or is it just the name of the cron job?

It's the latter: Only visible within the code and not visible in the web UI for end users in any way 👍 

It helps devs when looking at this file (after some time or when landing here for the first time) to quickly grasp how often each job runs, instead of having to parse the cron syntax in their (our) mind or having to use a tool for this 😄

File: app/queries/comments/community_wellness_query.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                          SELECT count(id) AS number_of_comments,
+                                 user_id AS comment_counts_user_id
+                          FROM comments
+                          WHERE created_at >= (now() - interval '7' day)
+                          GROUP BY user_id) AS comment_counts
+                          ON comments.user_id = comment_counts_user_id
+                          AND comment_counts.number_of_comments > 1
+                                /* Don’t select anything older than 224 days ago, or 32 weeks ago */
+                          WHERE created_at > (now() - interval '224' day)
+                          AND created_at < (now() - interval '3' day)
```
Comment (fdocr) [inline app/queries/comments/community_wellness_query.rb:64]: The query now filters comments so they're at least 3 days old before they count towards the badge. This is intended so mods have time to flag comments before the badge is awarded to users [as per this conversation](https://github.com/forem/forem/issues/17310#issuecomment-1110045254)

File: app/services/badges/award_community_wellness.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+          user.badge_achievements.create(
+            badge_id: badge_id,
+            rewarding_context_message_markdown: generate_message,
+          )
+        else
+          # If FeatureFlag isn't enabled only track which users would've
+          # received the badge to get an understanding of the service would work
+          Ahoy.instance&.track("Community Wellness Badge Award", user_id: user.id, week: @week_streak)
+        end
```
Comment (fdocr) [inline app/services/badges/award_community_wellness.rb]: The idea behind the Feature Flag here is to gather data of how the Badge awarding works using Ahoy on real instances before actually giving them to users.

File: config/schedule.yml [hunk 7]
```diff
@@ -78,6 +78,14 @@ award_contributor_badges_from_github:
     - ""
     - award_contributor_from_github
     - ""
+award_community_wellness_badges:
+  description: "Awards 'Community Wellness' badges to users (runs daily at 00:45 UTC)"
+  cron: "45 0 * * *"
```
Comment (fdocr) [inline config/schedule.yml:83]: Some other jobs run at ~00:00 UTC so I chose minute 45 to reduce the load on that hour

File: app/services/badges/award_community_wellness.rb [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+          user.badge_achievements.create(
+            badge_id: badge_id,
+            rewarding_context_message_markdown: generate_message,
+          )
+        else
+          # If the FeatureFlag isn't enabled only track which users would get
+          # the badge to get an understanding of how the service will work
+          Ahoy.instance&.track("Community Wellness Badge Award", user_id: user.id, week: @week_streak)
+        end
```
Comment (fdocr) [inline app/services/badges/award_community_wellness.rb:55]: The Feature Flag approach here allows us to run the service on production without awarding the badge. This should give us real life data on how the service executes and whether or not we need to tweak anything (performance, logic, etc).

File: spec/services/badges/award_community_wellness_spec.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+      allow(FeatureFlag).to receive(:enabled?).with(:community_wellness_badge).and_return(true)
+
+      expect do
+        described_class.call
+        users.each_with_index do |user, index|
+          # Each user must be tested against it's corresponding streak
+          expected_user_streak = reward_weeks[index]
+          badge_slug = "#{expected_user_streak}-week-community-wellness-streak"
+          expect(user.reload.badges.last.slug).to eq(badge_slug)
+        end
```
Comment (fdocr) [inline spec/services/badges/award_community_wellness_spec.rb:46]: Each mock user should be awarded one of the reward levels. I hope this spec isn't overly complex as I attempted to test all reward levels without too much redundancy with lots of similar `it` specs

File: app/services/badges/award_community_wellness.rb [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
```
Comment (fdocr) [inline app/services/badges/award_community_wellness.rb:54]: The Feature Flag approach here allows us to run the service on production without awarding the badge. This should give us real life data on how the service executes and whether or not we need to tweak anything (performance, logic, etc).

**Review by joshpuetz — APPROVED** [review]
Some questions inline, but nothing blocking. Well done!

File: app/queries/comments/community_wellness_query.rb [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+# result set, so we don't get any added value from instantiating AR objects
+module Comments
+  class CommunityWellnessQuery
+    def self.call
+      ActiveRecord::Base.connection.execute(sql_query).to_a
+    end
+
+    def self.sql_query
+      # Based on the following Blazer query:
+      # https://dev.to/admin/blazer/queries/677
```
[the remark this answers is not in the document]
Comment (fdocr) [reply app/queries/comments/community_wellness_query.rb:26]: I can absolutely agree with this! 

I'll start to work on internal documentation and discuss with Success team about updating https://admin.forem.com/docs/forem-basics/badges

File: config/schedule.yml (hunk 7 above)
Comment (fdocr) [reply config/schedule.yml:83]: It runs in less than 2 seconds. We checked the original query (this one has been slightly tweaked since) with @jgaskins and @jeremyf and looked good for a _"once or twice a day cadence"_ since it was doing well on index use despite having few unique values on `reaction.category`, if I remember correctly 😄 

1. Here is the [up-to-date Blazer query](https://dev.to/admin/blazer/queries/683)
2. Here is the [up-to-date EXPLAIN ANALYZE of the Blazer query](https://dev.to/admin/blazer/queries/684)
   - Which I can't myself read too well, but I think having `loops=1` throughout may be a good thing (?) 😅 

Note: Using these links for the docs

File: app/services/badges/award_community_wellness.rb [hunk 12]
```diff
@@ -0,0 +1,67 @@
+module Badges
+  class AwardCommunityWellness
+    REWARD_STREAK_WEEKS = [1, 2, 4, 8, 16, 32].freeze
+
+    def self.call
+      new.call
```
[the remark this answers is not in the document]
Comment (fdocr) [reply app/services/badges/award_community_wellness.rb:6]: I started off from the `def self.call(...)` style and never noticed this when cleaning up. I was aiming for consistency but it's not exactly that use case here since we don't benefit from the initializer and there are other services that are implemented directly in the static method.

Conclusion: Great input here, I'm pushing the changes in a bit. Thanks!

File: app/services/badges/award_community_wellness.rb [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        #    - [1,4,17]
+        # We only care for active streak (starting at 1) so we need to filter
+        # these to check how far back the (continuous) streak goes
+        week_streak = 0
+        weeks_ago.each_with_index do |week, index|
+          # Must have 2 or more non-flagged comments posted on that week
+          next unless comment_counts[index] > 1
+
+          # Must be a consecutive streak
+          next unless week_streak + 1 == week
```
Comment (jeremyf) [inline app/services/badges/award_community_wellness.rb:27]: Good code comment, I scanned this once and knew it should be there but didn't quite process what it did.
